### PR TITLE
Add abstract_charm.py

### PR DESCRIPTION
**Pull request by carlcsaposs-canonical**
Share most of charm.py between kubernetes & machine charm

Depends on https://github.com/canonical/mysql-router-k8s-operator/pull/93

### CODE REVIEW

File: src/kubernetes_charm.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        return rock.Rock(unit=self.unit)
+
+    @property
+    def model_service_domain(self) -> str:
+        """K8s service domain for Juju model"""
+        # Example: "mysql-router-k8s-0.mysql-router-k8s-endpoints.my-model.svc.cluster.local"
+        fqdn = socket.getfqdn()
+        # Example: "mysql-router-k8s-0.mysql-router-k8s-endpoints."
+        prefix = f"{self.unit.name.replace('/', '-')}.{self.app.name}-endpoints."
+        assert fqdn.startswith(f"{prefix}{self.model.name}.")
```
Comment (paulomach) [inline src/kubernetes_charm.py:55]: any special reason to make the assertion?

File: charmcraft.yaml [hunk 2]
```diff
@@ -11,6 +11,7 @@ bases:
         channel: "22.04"
 parts:
   charm:
+    charm-entrypoint: src/kubernetes_charm.py
```
Comment (paulomach) [inline charmcraft.yaml:14]: this is handy

File: src/kubernetes_charm.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        # Example: my-model.svc.cluster.local
+        return fqdn.removeprefix(prefix)
+
+    @property
+    def _host(self) -> str:
+        """K8s hostname for MySQL Router"""
```
Comment (paulomach) [inline src/kubernetes_charm.py:61]:
```suggestion
        """K8s service hostname for MySQL Router"""
```

**Review by paulomach — APPROVED** [review]
ignoreable nitpick, otherwise lgtm

**Review by shayancanonical — APPROVED** [review]

**Review by taurus-forever — APPROVED** [review]
Nice!

**Pull request merged**